index.js

```diff
@@ -16,76 +17,93 @@ function genKey(password, salt) {
     });
   });
 }
-function cryptoInit(password, modP) {
+function cryptoInit(password, options) {
   var db = this;
   var key, pub;
   var turnedOff = false;
-  return db.get(configId).catch(function (err) {
+  var ignore = ['_id', '_rev']
+
+  if (options && options.ignore) {
+    ignore = ignore.concat(options.ignore)
+  }
+  
+  var pending = db.get(configId).then(function (doc){
+    if (!doc.salt) {
+      throw {
+        status: 'invalid',
+        doc: doc
+      };
+    }
+    return doc;
+  }).catch(function (err) {
+    var doc;
     if (err.status === 404) {
-      var doc = {
+      doc = {
         _id: configId,
-        salt: crypto.randomBytes(16).toString('hex')
+        salt: randomBytes(16).toString('hex')
       };
+    } else if (err.status === 'invalid' && err.doc) {
+      doc = err.doc;
+      doc.salt = randomBytes(16).toString('hex');
+    }
+    if (doc) {
       return db.put(doc).then(function () {
         return doc;
       });
     }
     throw err;
   }).then(function (doc) {
-    var dh;
-    if (typeof modP === 'string') {
-      dh = crypto.getDiffieHellman(modP);
-      dh.generateKeys();
-      pub = dh.getPublicKey();
-      password = dh.computeSecret(password);
-    } else if (Buffer.isBuffer(modP)) {
-      dh = crypto.createDiffieHellman(modP);
-      dh.generateKeys();
-      pub = dh.getPublicKey();
-      password = dh.computeSecret(password);
-    }
     return genKey(password, new Buffer(doc.salt, 'hex'));
   }).then(function (_key) {
     password = null;
-    key = _key;
-    db.transform({
-      incoming: encrypt,
-      outgoing: decrypt
-    });
-    db.removeCrypto = function () {
+    if (turnedOff) {
       randomize(key);
-      turnedOff = true;
-    };
-    if (pub) {
-      return pub;
+    } else {
+      key = _key;
     }
   });
-  function encrypt(doc) {
-    if (turnedOff) {
-      return doc;
-    }
-    var id, rev;
-    if ('_id' in doc) {
-      id = doc._id;
-      delete doc._id;
-    } else {
-      id = uuid.v4();
+  db.transform({
+    incoming: function (doc) {
+      return pending.then(function () {
+        return encrypt(doc);
+      });
+    },
+    outgoing: function (doc) {
+      return pending.then(function () {
+        return decrypt(doc);
+      });
     }
-    if ('_rev' in doc) {
-      rev = doc._rev;
-      delete doc._rev;
+  });
+  db.removeCrypto = function () {
+    if (key) {
+      randomize(key);
     }
-    var nonce = crypto.randomBytes(12);
-    var data = JSON.stringify(doc);
+    turnedOff = true;
+  };
+
+  function encrypt(doc) {
+    var nonce = randomBytes(12)
     var outDoc = {
-      _id: id,
       nonce: nonce.toString('hex')
     };
-    if (rev) {
-      outDoc._rev = rev;
+    // for loop performs better than .forEach etc
+    for (var i = 0, len = ignore.length; i < len; i++) {
+      outDoc[ignore[i]] = doc[ignore[i]]
+      delete doc[ignore[i]]
```

Comment:
`doc` is the object that gets stringifyied and encrypted into `outDoc.data`:

```js
var data = JSON.stringify(doc);
var cipher = chacha.createCipher(key, nonce);
cipher.setAAD(new Buffer(outDoc._id));
outDoc.data = cipher.update(data).toString('hex');
```

All properties that shall be ignored are deleted from `doc` and directly set on `outDoc`, so that they will remain as unencrypted properties, while `.data` has the encrypted bits.

In the `decrypt` method we do the same in reverse, after reading out all the encrypted properties, we directly copy the unencrypted properties from `doc` to `out`